assemblers/org.osbuild.qcow2

```diff
@@ -45,36 +38,62 @@ def main(tree, output_dir, options, loop_client):
     if size % 512 != 0:
         raise ValueError("`size` must be a multiple of sector size (512)")
 
-    image = f"/tmp/osbuild-image.raw"
-    mountpoint = f"/tmp/osbuild-mnt"
+    image = "/tmp/osbuild-image.raw"
+    grub2_core = "/tmp/grub2-core.img"
+    mountpoint = "/tmp/osbuild-mnt"
 
     # Create an empty image file
     subprocess.run(["truncate", "--size", str(size), image], check=True)
 
     # Set up the partition table of the image
     partition_table = "label: mbr\nbootable, type=83"
     subprocess.run(["sfdisk", "-q", image], input=partition_table, encoding='utf-8', check=True)
+
     r = subprocess.run(["sfdisk", "--json", image], stdout=subprocess.PIPE, encoding='utf-8', check=True)
     partition_table = json.loads(r.stdout)
     partition = partition_table["partitiontable"]["partitions"][0]
     partition_offset = partition["start"] * 512
     partition_size = partition["size"] * 512
 
-    # Populate the first partition of the image with an ext4 fs and fill it with the contents of the
-    # tree we are operating on.
+    # Install the level-1 bootloader into the start of the MBR
+    # The purpose of this is simply to jump into the lever-2 bootloader.
+    with loop_device(loop_client, image, 512) as mbr, \
+        open(mbr, "wb") as mbr_f, \
+        open("/usr/lib/grub/i386-pc/boot.img", "rb") as boot_f:
+        # We must map a full 512 sector of the image, and the boot.img file is
+        # also 512 bytes, but we must only copy the first 440 bytes, as these
+        # contain the bootstrapping code. The rest of the first sector contains
+        # the partition table, and must not be overwritten.
+        mbr_f.write(boot_f.read()[:440])
+
+    # Create the level-2 bootloader
+    # The purpose of this is to find the grub modules and configuration
+    # to be able to start the level-3 bootloader. It contains the modules
+    # necessary to do this, but nothing else.
+    subprocess.run(["grub2-mkimage",
+                    "--directory", "/usr/lib/grub/i386-pc",
+                    "--prefix", "(,msdos1)/boot/grub2",
+                    "--format", "i386-pc",
+                    "--compression", "auto",
+                    "--output", grub2_core,
+                    "ext2", "part_msdos", "biosdisk"],
+                   check=True)
```

Comment:
If you hardcode the prefix then I think that this may be executed in the `org.osbuild.grub2` stage after copying the GRUB modules. The only reason that you had in the assembler was to specify the UUID in the embedded config file which wasn't known at the time the OS file system tree is assembled.

Reply:
Actually, the reason I did this in the assembler was just that I realised there was no need to store these files to /boot at all, might as well just generate them on the fly and insert them directly into the image.

Reply:
Right, makes sense to me.